### PR TITLE
support joint matrix prefetch

### PR DESCRIPTION
**Pull request by VyacheslavLevytskyy**
This PR aims to introduce CooperativeMatrixPrefetchINTEL capability and operation, and make initial introduction of entities in llvm-spirv translator.

### CODE REVIEW

**Comment by MrSidims** [comment]
>> Failed Tests (1):
  LLVM_SPIRV :: lshr-constexpr.ll
  
  not related to the patch

File: test/extensions/INTEL/SPV_INTEL_joint_matrix/cooperative_matrix.ll [hunk 1]
```diff
@@ -0,0 +1,171 @@
+; This is an adapted copy of test/extensions/KHR/SPV_KHR_cooperative_matrix/cooperative_matrix.ll
```
Comment (MrSidims) [inline test/extensions/INTEL/SPV_INTEL_joint_matrix/cooperative_matrix.ll:1]: Lets name the test appropriately, like cooperative_matrix_prefetch.ll

File: test/extensions/INTEL/SPV_INTEL_joint_matrix/cooperative_matrix.ll [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+; CHECK-LLVM: call spir_func i32 @_Z34__spirv_CooperativeMatrixLengthKHRPU3AS144__spirv_CooperativeMatrixKHR__char_0_12_48_3(target("spirv.CooperativeMatrixKHR", i8, 0, 12, 48, 3)
+; CHECK-LLVM: call spir_func target("spirv.CooperativeMatrixKHR", i8, 2, 48, 12, 3) @_Z86__spirv_CooperativeMatrixLoadKHR_RPU3AS144__spirv_CooperativeMatrixKHR__char_2_48_12_3PU3AS4cl
+; CHECK-LLVM: call spir_func target("spirv.CooperativeMatrixKHR", i32, 3, 12, 12, 3) @_Z34__spirv_CooperativeMatrixMulAddKHRPU3AS144__spirv_CooperativeMatrixKHR__char_0_12_48_3PU3AS144__spirv_CooperativeMatrixKHR__char_2_48_12_3PU3AS144__spirv_CooperativeMatrixKHR__uint_3_12_12_3i(target("spirv.CooperativeMatrixKHR", i8, 0, 12, 48, 3) %{{.*}}, target("spirv.CooperativeMatrixKHR", i8, 2, 48, 12, 3) %{{.*}}, target("spirv.CooperativeMatrixKHR", i32, 3, 12, 12, 3)
+; CHECK-LLVM: call spir_func void @_Z33__spirv_CooperativeMatrixStoreKHRPU3AS4iPU3AS144__spirv_CooperativeMatrixKHR__uint_3_12_12_3ili(ptr addrspace(4) %call.ascast.i.i, target("spirv.CooperativeMatrixKHR", i32, 3, 12, 12, 3)
```
Comment (MrSidims) [inline test/extensions/INTEL/SPV_INTEL_joint_matrix/cooperative_matrix.ll:42]:
```suggestion
; CHECK-LLVM: call spir_func void @_Z33__spirv_CooperativeMatrixStoreKHRPU3AS4iPU3AS144__spirv_CooperativeMatrixKHR__uint_3_12_12_3ili(ptr addrspace(4) %{{.*}}, target("spirv.CooperativeMatrixKHR", i32, 3, 12, 12, 3)
```

File: test/extensions/INTEL/SPV_INTEL_joint_matrix/cooperative_matrix.ll [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+; TODO: Pass Matrix Type Id instead of Matrix Id to CooperativeMatrixLengthKHR.
+; CHECK-SPIRV: CooperativeMatrixLengthKHR [[#Int32Ty]] [[#]] [[#Load1]]
+; CHECK-SPIRV: CooperativeMatrixPrefetchINTEL
+; CHECK-SPIRV: CooperativeMatrixLoadKHR [[#MatTy3]]
+; CHECK-SPIRV: CooperativeMatrixMulAddKHR [[#MatTy1]]
+; CHECK-SPIRV: CooperativeMatrixStoreKHR
+
+
+; CHECK-LLVM: call spir_func target("spirv.CooperativeMatrixKHR", i32, 3, 12, 12, 3) @_Z26__spirv_CompositeConstructi(i32 0)
+; CHECK-LLVM: call spir_func void @_Z38__spirv_CooperativeMatrixPrefetchINTELPU3AS4ciiiiil(ptr addrspace(4) %call.ascast.i66.i, i32 0, i32 0, i32 1, i32 1, i32 0, i64 %_arg_K)
```
Comment (MrSidims) [inline test/extensions/INTEL/SPV_INTEL_joint_matrix/cooperative_matrix.ll:37]: It's worth to check that the same pointer is being passed to prefetch and load, so:
```suggestion
; CHECK-LLVM: call spir_func void @_Z38__spirv_CooperativeMatrixPrefetchINTELPU3AS4ciiiiil(ptr addrspace(4) %[[MatrixPtr:[%0-9a-z.]+]], i32 0, i32 0, i32 1, i32 1, i32 0, i64 %_arg_K)
```

File: test/extensions/INTEL/SPV_INTEL_joint_matrix/cooperative_matrix.ll [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+; CHECK-LLVM: call spir_func target("spirv.CooperativeMatrixKHR", i8, 0, 12, 48, 3) @_Z86__spirv_CooperativeMatrixLoadKHR_RPU3AS144__spirv_CooperativeMatrixKHR__char_0_12_48_3PU3AS4clii
```
Comment (MrSidims) [inline test/extensions/INTEL/SPV_INTEL_joint_matrix/cooperative_matrix.ll:38]: And check for %[[MatrixPtr]] parameter here

File: lib/SPIRV/libSPIRV/SPIRVInstruction.h [hunk 5]
```diff
@@ -314,9 +314,7 @@ class SPIRVInstTemplateBase : public SPIRVInstruction {
     return Operands;
   }
 
-  virtual SPIRVValue *getOperand(unsigned I) {
-    return getOpValue(I);
-  }
+  virtual SPIRVValue *getOperand(unsigned I) { return getOpValue(I); }
```
Comment (MrSidims) [inline lib/SPIRV/libSPIRV/SPIRVInstruction.h:317]: better not to add non-related changes to the patch

File: lib/SPIRV/libSPIRV/SPIRVInstruction.h [hunk 6]
```diff
@@ -400,12 +398,12 @@ class SPIRVMemoryAccess {
     }
     if (MemoryAccess[0] & MemoryAccessAliasScopeINTELMaskMask) {
       assert(MemoryAccess.size() > MemAccessNumParam &&
-          "Aliasing operand is missing");
+             "Aliasing operand is missing");
```
Comment (MrSidims) [inline lib/SPIRV/libSPIRV/SPIRVInstruction.h:401]: same

File: lib/SPIRV/libSPIRV/spirv_internal.hpp [hunk 7]
```diff
@@ -52,10 +52,7 @@ enum InternalSourceLanguageNonSemanticDI {
   ISourceLanguageCPP20 = 219,
 };
 
-enum InternalLinkageType {
-  ILTPrev = LinkageTypeMax - 2,
-  ILTInternal
-};
+enum InternalLinkageType { ILTPrev = LinkageTypeMax - 2, ILTInternal };
```
Comment (MrSidims) [inline lib/SPIRV/libSPIRV/spirv_internal.hpp:55]: While this suggesting comes from clang-format, probably enums should be better to be split into multiple lines. I'll add ignore format to this file I guess

File: lib/SPIRV/libSPIRV/SPIRVInstruction.h (hunk 5 above)
Comment (VyacheslavLevytskyy) [reply lib/SPIRV/libSPIRV/SPIRVInstruction.h:317]: agree, this and following are clang-formats artifacts, I'll revert to the previous state

**Comment by VyacheslavLevytskyy** [comment]
@MrSidims thanks for suggestions; can you please check them applied

**Comment by MrSidims** [comment]
clang-format is unrelated (and seems like something is no longer working, as it no longer is printing the patch file that should be applied, will take a look tomorrow).

**Review by MrSidims — APPROVED** [review]

**Pull request merged**